contracts/RelayHub.sol

```diff
@@ -205,85 +238,168 @@ contract RelayHub is IRelayHub {
-        // We should advance the nonce here, as once we get to this point, the recipient pays for the transaction whether if the relayed call is reverted or not.
+        // From this point on, this transaction will not revert nor run out of gas, and the recipient will be charged
+        // for the gas spent.
+
+        // The sender's nonce is advanced to prevent transaction replays.
         nonces[from]++;
-        RelayCallStatus status = RelayCallStatus.OK;
-        if (!successPrePost) {
-            status = RelayCallStatus.PostRelayedFailed;
-        } else if (LibBytes.readUint256(relayedCallSuccess, 0) == 2) {
-            status = RelayCallStatus.PreRelayedFailed;
-        } else if (LibBytes.readUint256(relayedCallSuccess, 0) == 0) {
-            status = RelayCallStatus.RelayedCallFailed;
-        }
-        // Relay transactionFee is in %.  E.g. if transactionFee=40, payment will be 1.4*usedGas.
-        uint charge = (gasOverhead + initialGas - gasleft()) * gasPrice * (100 + transactionFee) / 100;
-        emitTransactionRelayed(msg.sender, from, to, encodedFunction, uint(status), charge);
-        // We already checked at the beginning that the recipient has enough balance. This is more of a sanity check/safeMath before we substract from balance
-        require(balances[to] >= charge, "Should not get here");
-        balances[to] -= charge;
+
+        // Calls to the recipient are performed atomically inside an inner transaction which may revert in case of
+        // errors in the recipient. In either case (revert or regular execution) the return data encodes the
+        // RelayCallStatus value.
+        (, bytes memory relayCallStatus) = address(this).call(abi.encodeWithSelector(this.recipientCallsAtomic.selector, from, recipient, msg.sender, encodedFunction, transactionFee, gasLimit, initialGas));
+        RelayCallStatus status = abi.decode(relayCallStatus, (RelayCallStatus));
+
+        // Regardless of the outcome of the relayed transaction, the recipient is now charged.
+        uint256 charge = getChargedAmount(gasOverhead + initialGas - gasleft(), gasPrice, transactionFee);
+
+        // We've already checked that the recipient has enough balance to pay for the relayed transaction, this is only
+        // a sanity check to prevent overflows in case of bugs.msg.sender
+        require(balances[recipient] >= charge, "Should not get here");
+        balances[recipient] -= charge;
         balances[relays[msg.sender].owner] += charge;
-    }
 
-    // Bypassing "stack too deep"... in relayCall()
-    function emitTransactionRelayed(address sender, address from, address to, bytes memory encodedFunction, uint status, uint charge) internal {
-        emit TransactionRelayed(sender, from, to, LibBytes.readBytes4(encodedFunction, 0), status, charge);
+        emit TransactionRelayed(msg.sender, from, recipient, abi.decode(encodedFunction, (bytes4)), uint256(status), charge);
```

Comment:
We said that `decode( ..bytes4)` is dangerous. So we do need `readBytes4`, and thus re-introduce `emitTransactionRelayed`